### PR TITLE
Don't use __getstate__ for pickling for objects with a custom __reduce__.

### PR DESCRIPTION
**Pull request by robertwb**
This fixes https://github.com/apache/beam/pull/33137

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 1]
```diff
@@ -500,7 +500,9 @@ def encode_special_deterministic(self, value, stream):
         self.encode_to_stream(value.value, stream, True)
       except Exception as e:
         raise TypeError(self._deterministic_encoding_error_msg(value)) from e
-    elif hasattr(value, "__getstate__"):
+    elif (hasattr(value, "__getstate__") and
+          # https://github.com/apache/beam/issues/33020
+          type(value).__reduce__ == object.__reduce):
       if not hasattr(value, "__setstate__"):
```
Comment (liferoad) [inline sdks/python/apache_beam/coders/coder_impl.py:506]: are we sure this works?
```
class MyClass:
    def __init__(self, value):
        self.value = value

    def __reduce__(self):
        return (self.__class__, (self.value,))

tt = MyClass(1)
print(object.__reduce__ == type(tt).__reduce__)
```
This also returns False. I might misunderstand how this works.

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 2]
```diff
@@ -500,7 +500,9 @@ def encode_special_deterministic(self, value, stream):
         self.encode_to_stream(value.value, stream, True)
       except Exception as e:
         raise TypeError(self._deterministic_encoding_error_msg(value)) from e
-    elif hasattr(value, "__getstate__"):
+    elif (hasattr(value, "__getstate__") and
+          # https://github.com/apache/beam/issues/33020
+          type(value).__reduce__ == object.__reduce):
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/coders/coder_impl.py]: Yes. Fixed.

File: sdks/python/apache_beam/coders/coder_impl.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/coders/coder_impl.py:506]: We want this to return `False` as MyClass has a custom `__reduce__` method (meaning relying on `__set/getstate__` is not safe).

**Review by liferoad — APPROVED** [review]

File: sdks/python/apache_beam/coders/coder_impl.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/coders/coder_impl.py:506]: Well, thanks goes to you for hunting down the root cause; it was not obvious at all what was going on here :). I'll merge once tests pass.

**Pull request merged**